### PR TITLE
Initial working conversion

### PR DESCRIPTION
**Pull request by chbatey**

### CODE REVIEW

**Comment by patriknw** [comment]
I missed that this PR was over here. Placing in my todo but will not make it today.

File: src/main/resources/logback.xml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<configuration>
+
+    <!-- We define all our logging task into the appender like CONSOLE, FILE etc.
+     Here we are logging the ConsoleAppender, as the name indicates, appends on the console.  -->
+    <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
+        <encoder class="ch.qos.logback.classic.encoder.PatternLayoutEncoder">
+            <Pattern>%d{HH:mm:ss.SSS} [%thread] %-5level %logger{36} - %msg%n
+            </Pattern>
+        </encoder>
+    </appender>
```
Comment (patriknw) [inline src/main/resources/logback.xml:8]: use the async appender as in the logging.md docs

File: src/main/resources/application.conf [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    # Seed nodes are a way to have a node join the cluster (or form a new cluster) from configuration.
+    seed-nodes = [
+      "akka://ClusterSystem@127.0.0.1:2551",
+      "akka://ClusterSystem@127.0.0.1:2552",
+      "akka://ClusterSystem@127.0.0.1:2553",
+      "akka://ClusterSystem@127.0.0.1:2554"]
+
+    # Only for convenience in the quickstart, auto-downing should not be used for actual applications.
+    # Read more here: http://doc.akka.io/docs/akka/current/scala/cluster-usage.html#auto-downing-do-not-use-
+    auto-down-unreachable-after = 10s
```
Comment (patriknw) [inline src/main/resources/application.conf]: remove

File: src/main/resources/application.conf [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  actor {
+    # Must be set like this to use Akka Cluster
+    provider = cluster
+
+    serialization-bindings {
+      "worker.CborSerializable" = jackson-cbor
+    }
+  }
+
+ loglevel = INFO
```
Comment (patriknw) [inline src/main/resources/application.conf]: DEBUG, and filter with logback

File: src/main/scala/worker/Worker.scala [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    .getDuration("distributed-workers.worker-registration-interval")
+    .toNanos
+    .nano
+
+  private val workAckTimeout = ctx.system.settings.config
+    .getDuration("distributed-workers.work-ack-timeout")
+    .toNanos
+    .nano
+
+  timers.startTimerAtFixedRate("register", Register, registerInterval)
```
Comment (patriknw) [inline src/main/scala/worker/Worker.scala]: prefer startTimerWithFixedDelay

File: src/main/scala/worker/MasterSingleton.scala [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import akka.cluster.typed._
+import worker.Master.MasterCommand
+
+object MasterSingleton {
+
+  private val singletonName = "master"
+  private val singletonRole = "back-end"
+
+  // #singleton
+  def singleton(system: ActorSystem[_]): ActorRef[MasterCommand] = {
```
Comment (patriknw) [inline src/main/scala/worker/MasterSingleton.scala]: name this init?

File: src/main/scala/worker/Master.scala [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+  def apply(workTimeout: FiniteDuration): Behavior[MasterCommand] =
+    Behaviors.setup { ctx =>
+      Behaviors.withTimers { timers =>
+        timers.startSingleTimer("cleanup", Cleanup, workTimeout / 2)
+
+        // No typed pub sub yet
+        val mediator = DistributedPubSub(ctx.system.toClassic).mediator
+
+        // the set of available workers is not event sourced as it depends on the current set of workers
+        var workers = Map[String, WorkerState]()
```
Comment (patriknw) [inline src/main/scala/worker/Master.scala:56]: would it be possible to split this into two actors? one managing the workers and another for the event sourced WorkState? 

They could still be tightly coupled by sharing same messages.

File: src/main/scala/worker/Main.scala [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  /**
+    * Start a front end node that will submit work to the backend nodes
+    */
+  // #front-end
+  def startFrontEnd(port: Int): Unit = {
+    ActorSystem[Nothing](Behaviors.setup[Nothing] { ctx =>
+      ctx.spawn(FrontEnd(), "front-end")
+      ctx.spawn(WorkResultConsumer(), "consumer")
```
Comment (patriknw) [inline src/main/scala/worker/Main.scala]: I tend to like having one single guardian Behavior, which looks at the cluster roles to start different things. This also works, not sure what is best here?

**Review by patriknw — COMMENTED** [review]
looking good (reviewed this yesterday)

File: src/main/resources/application.conf (hunk 2 above)
Comment (chbatey) [reply src/main/resources/application.conf]: Oops!

File: src/main/scala/worker/Main.scala (hunk 7 above)
Comment (chbatey) [reply src/main/scala/worker/Main.scala]: I think your way is more common, I'll switch it

File: src/main/resources/logback.xml (hunk 1 above)
Comment (chbatey) [reply src/main/resources/logback.xml:8]: I'll not do that for the console one

File: src/main/scala/worker/Master.scala (hunk 6 above)
Comment (chbatey) [reply src/main/scala/worker/Master.scala:56]: Yes I think it is worth trying a different pattern here, the var is easy to get wrong

Comment (chbatey) [reply src/main/scala/worker/Master.scala:56]: It does get very messy unless we were to change it entirely. As we often update the persistent state and the transient state at the same time

Comment (patriknw) [reply src/main/scala/worker/Master.scala:56]: Ok, leave it as is then

**Pull request closed without merging**